simulator/src/main/java/peersim/kademlia/das/operations/SamplingOperation.java

```diff
@@ -92,70 +106,111 @@ public BigInteger getRadiusNonValidator() {
+              nodes.put(id, new Node(id));
+              nodes.get(id).addSample(samples.get(sample));
+            } else {
+              nodes.get(id).addSample(samples.get(sample));
+            }
           }
-          if (nonValidatorsBySample != null && nonValidatorsBySample.size() > 0) {
-            nodes.addAll(nonValidatorsBySample);
-            found = true;
+          found = true;
+        }
+        if (nonValidatorsBySample != null && nonValidatorsBySample.size() > 0) {
+          for (BigInteger id : nonValidatorsBySample) {
+            if (!nodes.containsKey(id)) {
+              nodes.put(id, new Node(id));
+              nodes.get(id).addSample(samples.get(sample));
+            } else {
+              nodes.get(id).addSample(samples.get(sample));
+            }
           }
-          if (!found && callback != null) callback.missing(sample, this);
+          found = true;
         }
+
+        if (!found && callback != null) callback.missing(sample, this);
       }
-      // Collections.shuffle(nodes);
+    }
+  }
+
+  public BigInteger[] doSampling() {
 
-      for (BigInteger node : nodes) {
-        if (closestSet.get(node) == null) {
-          closestSet.put(node, true);
+    aggressiveness += KademliaCommonConfigDas.aggressiveness_step;
+    for (Node n : nodes.values()) n.setAgressiveness(aggressiveness);
+    //System.out.println(
+    //    this.srcNode + "] Do sampling " + aggressiveness + " " + this.getId() + " " + nodes.size());
+    // orderNodes();
+    List<BigInteger> result = new ArrayList<>();
+    for (Node n : nodes.values()) {
+      /*System.out.println(
+          this.srcNode + "] Querying node " + n.getId() + " " + +n.getScore() + " " + this.getId());
+      for (FetchingSample fs : n.getSamples())
+        System.out.println(
+            this.srcNode + "] Sample " + fs.beingFetchedFrom.size() + " " + fs.isDownloaded());*/
+
+      if (!n.isBeingAsked() && n.getScore() > 0) { // break;
+        n.setBeingAsked(true);
+        this.available_requests++;
+        for (FetchingSample s : n.getSamples()) {
+          s.addFetchingNode(n);
         }
+        result.add(n.getId());
       }
     }
-    if (!closestSet.isEmpty()) {
-      res = closestSet.entrySet().iterator().next().getKey();
-      this.available_requests--; // decrease available requets
-      closestSet.remove(res);
-    }
-    return res;
+
+    return result.toArray(new BigInteger[0]);
   }
 
-  public boolean increaseRadius(int multiplier) {
+  /*public boolean increaseRadius(int multiplier) {
+    if (timesIncreased == KademliaCommonConfigDas.multiplyRadiusLimit) return false;
     radiusValidator = radiusValidator.multiply(BigInteger.valueOf(multiplier));
     if (Block.MAX_KEY.compareTo(radiusValidator) <= 0) {
       radiusValidator = Block.MAX_KEY;
       return false;
     }
+    createNodes();
     return true;
+  }*/
+
+  private void orderNodes() {
+    for (Node n : nodes.values()) n.setAgressiveness(aggressiveness);
```

Comment:
I don't get again why we need to loop through all the nodes to set aggressiveness

Reply:
orderNodes function is actually not used. i removed it from the code